File: Top_Interview_150/Array_String/python/6.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:

        if numRows == 1:
            return s

        n = len(s)
        result = [""] * numRows

        j = 0
        d = 1
        for i in range(n):
            result[j] += s[i]

            if j == 0:
                d = 1
            elif j == numRows - 1:
                d = -1
            j += d

        return "".join(result)


solution = Solution()
print(solution.convert(s="AB", numRows=1))

File: Top_Interview_150/Array_String/cpp/6.cpp
#include <string>
#include <vector>

using namespace std;

class Solution
{
public:
    string convert(string s, int numRows)
    {
        if (numRows == 1)
            return s;

        int n = s.size();
        vector<string> components(numRows, "");

        int j = 0;
        int d = 1;
        for (int i = 0; i < n; ++i)
        {
            components[j] += s[i];

            if (j == 0)
                d = 1;
            else if (j == numRows - 1)
                d = -1;
            j += d;
        }

        string result = "";
        for (int i = 0; i < numRows; ++i)
        {
            result += components[i];
        }

        return result;
    }
};